### ISSUE TITLE
Error after VS update 16.1.0

### ISSUE THREAD

**Issue by lvanvugt** [deadlydog/VS.DiffAllFiles#32]
After having installed VS update 16.1.0 _Pending Changes_ pane shows error:

```
Microsoft.VisualStudio.Composition.CompositionFailedException: An exception was thrown while initializing part "VS_DiffAllFiles.Sections.PendingChangesSection". ---> System.Windows.Markup.XamlParseException: Provide value on 'QuickConverter.MultiBinding' threw an exception. ---> System.Exception: Failed to tokenize expression "!$P0 || ($P0 ## $P1) ? Visibility.Visible : Visibility.Collapsed". Did you forget a '$'?
   at QuickConverter.EquationTokenizer.Tokenize(String expression)
   at QuickConverter.MultiBinding.GetLambda(String expression)
   at QuickConverter.MultiBinding.ProvideValue(IServiceProvider serviceProvider)
   at MS.Internal.Xaml.Runtime.ClrObjectRuntime.CallProvideValue(MarkupExtension me, IServiceProvider serviceProvider)
   --- End of inner exception stack trace ---
   at System.Windows.Markup.WpfXamlLoader.Load(XamlReader xamlReader, IXamlObjectWriterFactory writerFactory, Boolean skipJournaledProperties, Object rootObject, XamlObjectWriterSettings settings, Uri baseUri)
   at System.Windows.Markup.WpfXamlLoader.LoadBaml(XamlReader xamlReader, Boolean skipJournaledProperties, Object rootObject, XamlAccessLevel accessLevel, Uri baseUri)
   at System.Windows.Markup.XamlReader.LoadBaml(Stream stream, ParserContext parserContext, Object parent, Boolean closeStream)
   at System.Windows.Application.LoadComponent(Object component, Uri resourceLocator)
   at VS_DiffAllFiles.Sections.DiffAllFilesSectionControl.InitializeComponent() in C:\dev\Git\VS.DiffAllFiles\VS.DiffAllFiles.VS2019\obj\Release\DiffAllFilesSectionControl.xaml:line 1
   at VS_DiffAllFiles.DiffAllFilesBaseClasses.DiffAllFilesSectionBase..ctor() in C:\dev\Git\VS.DiffAllFiles\VS.DiffAllFiles\DiffAllFilesBaseClasses\DiffAllFilesSectionBase.cs:line 82
   at VS_DiffAllFiles.Sections.PendingChangesSection..ctor() in C:\dev\Git\VS.DiffAllFiles\VS.DiffAllFiles\Sections\PendingChangesSection.cs:line 27
   --- End of inner exception stack trace ---
   at Microsoft.VisualStudio.Composition.RuntimeExportProviderFactory.RuntimeExportProvider.RuntimePartLifecycleTracker.CreateValue()
   at Microsoft.VisualStudio.Composition.ExportProvider.PartLifecycleTracker.Create()
   at Microsoft.VisualStudio.Composition.ExportProvider.PartLifecycleTracker.MoveNext(PartLifecycleState nextState)
   at Microsoft.VisualStudio.Composition.ExportProvider.PartLifecycleTracker.MoveToState(PartLifecycleState requiredState)
   at Microsoft.VisualStudio.Composition.ExportProvider.PartLifecycleTracker.GetValueReadyToExpose()
   at Microsoft.VisualStudio.Composition.ExportProvider.<>c__DisplayClass53_0.<CreateExport>b__0()
   at Microsoft.VisualStudio.Composition.ExportProvider.<>c__DisplayClass55_0.<CreateExportFactoryExport>b__1()
   at Microsoft.VisualStudio.Composition.ExportProvider.<>c__DisplayClass54_0.<CreateExportFactory>b__0()
   at Microsoft.VisualStudio.Composition.DelegateServices.<>c__DisplayClass2_0`1.<As>b__0()
   at System.ComponentModel.Composition.ExportFactory`1.CreateExport()
   at Microsoft.VisualStudio.Composition.NetFxAdapters.MefV1ExportProvider.ComposablePartDefinitionForExportFactory.CreatePart()
   at System.ComponentModel.Composition.ReflectionModel.ExportFactoryCreator.LifetimeContext.GetExportLifetimeContextFromExport[T](Export export)
   at System.ComponentModel.Composition.ReflectionModel.ExportFactoryCreator.<>c__DisplayClass6_0`2.<CreateStronglyTypedExportFactoryOfTM>b__0()
   at System.ComponentModel.Composition.ExportFactory`1.CreateExport()
   at Microsoft.TeamFoundation.Controls.WPF.TeamExplorer.Framework.TeamExplorerSectionHost.Create()
```
**Note**
Not fully sure about the update version whether it is 16.1.0 exactly.

**Comment by deadlydog** [comment]
Hmmmm, that's odd. Just to confirm, you are seeing this issue on a TFVC repository? I've seen similar issues with Git repositories in the past, but not with TFVC.

**Comment by honfika** [comment]
@deadlydog I've the same problem, with Git.

**Comment by Invvard** [comment]
For the record, I've also this exception.
I don't know if it's related but it is impossible to install [your extension](https://marketplace.visualstudio.com/items?itemName=deadlydog.DiffAllFilesforVS2019) on VS2019 v16.1(.1)

Here is a screenshot of the error :
![error](https://user-images.githubusercontent.com/7305493/58348829-b2cf3f00-7e2f-11e9-882f-62d115ffef7c.png)

**Comment by deadlydog** [comment]
> @deadlydog I've the same problem, with Git.

@honfika are you getting this with VS 2019, or a different version of VS?

**Comment by honfika** [comment]
VS 2019 16.1.0
I'll try 16.1.1 in some minutes.

**Comment by deadlydog** [comment]
Thanks for opening new issue #33 for your problem @Invvard . We'll follow up with that problem in that issue 👍

**Comment by honfika** [comment]
Same problem in VS 2019 16.1.1 with Git

**Comment by deadlydog** [comment]
@honfika can you please post the full error message like the original poster did. I _think_ @ivanvugt's issue is in TFVC, not Git, so there may be 2 separate issues.

**Comment by honfika** [comment]
Microsoft.VisualStudio.Composition.CompositionFailedException: An exception was thrown while initializing part "VS_DiffAllFiles.Sections.GitChangesSection". ---> System.Windows.Markup.XamlParseException: Provide value on 'QuickConverter.MultiBinding' threw an exception. ---> System.Exception: Failed to tokenize expression "!$P0 || ($P0 ## $P1) ? Visibility.Visible : Visibility.Collapsed". Did you forget a '$'?
   at QuickConverter.EquationTokenizer.Tokenize(String expression)
   at QuickConverter.MultiBinding.GetLambda(String expression)
   at QuickConverter.MultiBinding.ProvideValue(IServiceProvider serviceProvider)
   at MS.Internal.Xaml.Runtime.ClrObjectRuntime.CallProvideValue(MarkupExtension me, IServiceProvider serviceProvider)
   --- End of inner exception stack trace ---
   at System.Windows.Markup.WpfXamlLoader.Load(XamlReader xamlReader, IXamlObjectWriterFactory writerFactory, Boolean skipJournaledProperties, Object rootObject, XamlObjectWriterSettings settings, Uri baseUri)
   at System.Windows.Markup.WpfXamlLoader.LoadBaml(XamlReader xamlReader, Boolean skipJournaledProperties, Object rootObject, XamlAccessLevel accessLevel, Uri baseUri)
   at System.Windows.Markup.XamlReader.LoadBaml(Stream stream, ParserContext parserContext, Object parent, Boolean closeStream)
   at System.Windows.Application.LoadComponent(Object component, Uri resourceLocator)
   at VS_DiffAllFiles.Sections.DiffAllFilesSectionControl.InitializeComponent() in C:\dev\Git\VS.DiffAllFiles\VS.DiffAllFiles.VS2019\obj\Release\DiffAllFilesSectionControl.xaml:line 1
   at VS_DiffAllFiles.DiffAllFilesBaseClasses.DiffAllFilesSectionBase..ctor() in C:\dev\Git\VS.DiffAllFiles\VS.DiffAllFiles\DiffAllFilesBaseClasses\DiffAllFilesSectionBase.cs:line 82
   at VS_DiffAllFiles.Sections.GitChangesSection..ctor() in C:\dev\Git\VS.DiffAllFiles\VS.DiffAllFiles\Sections\GitChangesSection.cs:line 154
   --- End of inner exception stack trace ---
   at Microsoft.VisualStudio.Composition.RuntimeExportProviderFactory.RuntimeExportProvider.RuntimePartLifecycleTracker.CreateValue()
   at Microsoft.VisualStudio.Composition.ExportProvider.PartLifecycleTracker.Create()
   at Microsoft.VisualStudio.Composition.ExportProvider.PartLifecycleTracker.MoveNext(PartLifecycleState nextState)
   at Microsoft.VisualStudio.Composition.ExportProvider.PartLifecycleTracker.MoveToState(PartLifecycleState requiredState)
   at Microsoft.VisualStudio.Composition.ExportProvider.PartLifecycleTracker.GetValueReadyToExpose()
   at Microsoft.VisualStudio.Composition.ExportProvider.<>c__DisplayClass53_0.<CreateExport>b__0()
   at Microsoft.VisualStudio.Composition.ExportProvider.<>c__DisplayClass55_0.<CreateExportFactoryExport>b__1()
   at Microsoft.VisualStudio.Composition.ExportProvider.<>c__DisplayClass54_0.<CreateExportFactory>b__0()
   at Microsoft.VisualStudio.Composition.DelegateServices.<>c__DisplayClass2_0`1.<As>b__0()
   at System.ComponentModel.Composition.ExportFactory`1.CreateExport()
   at Microsoft.VisualStudio.Composition.NetFxAdapters.MefV1ExportProvider.ComposablePartDefinitionForExportFactory.CreatePart()
   at System.ComponentModel.Composition.ReflectionModel.ExportFactoryCreator.LifetimeContext.GetExportLifetimeContextFromExport[T](Export export)
   at System.ComponentModel.Composition.ReflectionModel.ExportFactoryCreator.<>c__DisplayClass6_0`2.<CreateStronglyTypedExportFactoryOfTM>b__0()
   at System.ComponentModel.Composition.ExportFactory`1.CreateExport()
   at Microsoft.TeamFoundation.Controls.WPF.TeamExplorer.Framework.TeamExplorerSectionHost.Create()

**Comment by deadlydog** [comment]
Well the good news is that after updating to Visual Studio 16.1 I can reproduce the issue, both in the TFVC `Pending Changes` window and the Git `Changes` window. I'll try and find some time this weekend to dig into it some more. For now though, I've gotta get back to my day job :P

**Comment by lvanvugt** [comment]
> Hmmmm, that's odd. Just to confirm, you are seeing this issue on a TFVC repository? I've seen similar issues with Git repositories in the past, but not with TFVC.


TFVC

**Comment by xperiandri** [comment]
That's because you use synchronous load APIs which are deprecated.
Just update package to be async.
You can have a look at what to do by my commits in this extension
https://github.com/madskristensen/DialToolsForVS/commits/master

**Comment by deadlydog** [comment]
Yeah, it looks like in VS 2019 v16.0 they would just give a warning about synchronous extensions, but in v16.1 they disable them by default.

![image](https://user-images.githubusercontent.com/1187140/58386825-acbd9780-7fc2-11e9-90c7-aa6e435005a8.png)

If you click to `Allow synchronous autoload` then the extension works normally again.

I did have the extension using asynchronous loading, but it introduced issue #27 so I rolled that change back. Given that VS is now disabling synchronous extensions by default in v16.1+, I think I'll make it asynchronous again so it's usable by default, and look at fixing #27 later.

**Comment by Smurf-IV** [comment]
> Hmmmm, that's odd. Just to confirm, you are seeing this issue on a TFVC repository? I've seen similar issues with Git repositories in the past, but not with TFVC.

I Get this with TFS 2018 on premises - not git.

**Comment by deadlydog** [comment]
I've just released [v1.0.4 of the extension](https://marketplace.visualstudio.com/items?itemName=deadlydog.DiffAllFilesforVS2019). Give that a try and confirm it fixes the issue for you guys. Be aware though that #27 is still an issue now. Thanks again for all of the posts everyone :)

**Comment by lvanvugt** [comment]
> I've just released [v1.0.4 of the extension](https://marketplace.visualstudio.com/items?itemName=deadlydog.DiffAllFilesforVS2019). Give that a try and confirm it fixes the issue for you guys. Be aware though that #27 is still an issue now. Thanks again for all of the posts everyone :)

Updated the extension and loading fine now.Some checks showed it does the job op comparing right. Thanx!
